connection.go

```diff
@@ -0,0 +1,405 @@
+	}
+
+	err = ClientConnection.SubscribeUpdates()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func Close() {
+	ClientConnection.BrokerConnection.Close()
+}
+
+func Host(host string) Option {
+	return func(o *Options) error {
+		o.Host = host
+		return nil
+	}
+}
+
+func GetDefaultOptions() Options {
+	return Options{
+		Host: "broker.cost.memphis.dev",
+	}
+}
+
+func (c *Client) InitializeNatsConnection(token, host string) error {
+
+	splitedToken := strings.Split(token, ":::")
+	if len(splitedToken) != 2 {
+		return fmt.Errorf("memphis_kafka: not valid token")
+	}
+
+	JWT := splitedToken[0]
+	Nkey := splitedToken[1]
+
+	opts := []nats.Option{
+		nats.UserJWT(
+			func() (string, error) { // Callback to return the user JWT
+				return JWT, nil
+			},
+			func(nonce []byte) ([]byte, error) { // Callback to sign the nonce with user's NKey seed
+				userNKey, err := nkeys.FromSeed([]byte(Nkey))
+				if err != nil {
+					return nil, err
+				}
+				defer userNKey.Wipe()
+				return userNKey.Sign(nonce)
+			},
+		),
+		nats.ReconnectHandler(
+			func(nc *nats.Conn) {
+				natsConnectionID, err := c.generateNatsConnectionID()
+				if err != nil {
+					memphisKafkaErr("error reconnecting to memphis cost")
+				}
+
+				clientReconnectionUpdateReq := ClientReconnectionUpdateReq{
+					NewNatsClientID: natsConnectionID,
+					ClientID:        c.ClientID,
+				}
+
+				clientReconnectionUpdateReqBytes, err := json.Marshal(clientReconnectionUpdateReq)
+				if err != nil {
+					memphisKafkaErr("error reconnecting to memphis cost")
+				}
+
+				_, err = nc.Request(clientReconnectionUpdateSubject, clientReconnectionUpdateReqBytes, 5*time.Second)
+				if err != nil {
+					memphisKafkaErr("error reconnecting to memphis cost")
+				}
+
+				c.NatsClientID = natsConnectionID
+			},
+		),
+	}
+
+	nc, err := nats.Connect(host, opts...)
+	if err != nil {
+		return fmt.Errorf("memphis_kafka: error connecting to memphis cost")
+	}
+
+	js, err := nc.JetStream()
+	if err != nil {
+		return fmt.Errorf("memphis_kafka: error connecting to memphis cost")
+	}
+
+	c.BrokerConnection = nc
+	c.JSContext = js
+
+	natsConnectionID, err := c.generateNatsConnectionID()
+	if err != nil {
+		return fmt.Errorf("memphis_kafka: error connecting to memphis cost")
+	}
+	c.NatsClientID = natsConnectionID
+
+	return nil
+}
+
+func (c *Client) RegisterClient() error {
+	registerReq := RegisterReq{
+		NatsClientID: c.NatsClientID,
+		Language:     "go",
+	}
+
+	registerReqBytes, err := json.Marshal(registerReq)
+	if err != nil {
+		return fmt.Errorf("memphis_kafka: error registering client")
+	}
+
+	resp, err := c.BrokerConnection.Request(clientRegisterSubject, registerReqBytes, 5*time.Second)
+	if err != nil {
+		return fmt.Errorf("memphis_kafka: error registering client")
+	}
+
+	var registerResp RegisterResp
+	err = json.Unmarshal(resp.Data, &registerResp)
+	if err != nil {
+		return fmt.Errorf("memphis_kafka: error registering client")
+	}
+
+	if registerResp.Err != "" {
```

Comment:
I meant that max connection error should be returned from nats in this case not a custom message from the backend, so once this kind of error occurs you should raise this error